### ISSUE TITLE
Creating an SSH user fails with "Missing or invalid address for user creation mail"

### ISSUE THREAD

**Issue by jcameron** [virtualmin/virtualmin-gpl#768]
@iliaross looks like this is a side-effect of your recent changes..

The full stack trace is : 

![image](https://github.com/virtualmin/virtualmin-gpl/assets/102706/5405f1fa-7c92-4902-8f0d-4a96ad18ff47)

**Comment by jcameron** [comment]
Also, while you're in there, what is the purpose of this line : https://github.com/virtualmin/virtualmin-gpl/blob/master/edit_user.cgi#L170

Surely `1 ? "" : substr($user->{'home'}, length($d->{'home'})+1)` always evaluates to "" ?

**Comment by jcameron** [comment]
Also also, what's up with this code : 

`$d->{'mail'} ? undef : ['style="display: none"']`

Is it hiding some elements in the HTML? If so, why not just not display them?

**Comment by iliajie** [comment]
> Surely `1 ? "" : substr($user->{'home'}, length($d->{'home'})+1)` always evaluates to "" ?

Yes, correct. This is a harmless left-over. Fixed here https://github.com/virtualmin/virtualmin-gpl/commit/f767304ddacb95eb2b40972573c341f2c09cd1af &mdash; this code was derived from a main form, where a username is also expected to be edited.



> $d->{'mail'} ? undef : ['style="display: none"']

You're right, most of it was a left-over after I moved things around. The fix is here https://github.com/virtualmin/virtualmin-gpl/commit/b06b0e3e7a8b5c3f0b245575514b8846dc770c0b.


> looks like this is a side-effect of your recent changes.

Oh, this is a bug. The fix is here https://github.com/virtualmin/virtualmin-gpl/commit/ade636782fe47dd9e5da1beaf431e8b029da434c.

**Comment by jcameron** [comment]
Thanks for those fixes!

**Comment by iliajie** [comment]
You're welcome! If there is anything else, please let me know.

**Comment by jcameron** [comment]
I noticed another case where a hidden attribute is being used : https://github.com/virtualmin/virtualmin-gpl/blob/master/edit_user.cgi#L931

**Comment by iliajie** [comment]
Yeah .. What if a user which had mail enabled, later disables mail .. but still wants mail-related features to be preserved intact (when form saved without mail enabled)?

**Comment by jcameron** [comment]
We shoiuld just handle this in the save CGI, not by hiding fields in the HTML.

**Comment by iliajie** [comment]
Alright, I will review it then, and submit a fix ..

**Comment by iliajie** [comment]
> We shoiuld just handle this in the save CGI, not by hiding fields in the HTML.

Here is the fix https://github.com/virtualmin/virtualmin-gpl/commit/ba6564e8c8055ca299beb12bdf07c4bf727f9d40.

**Comment by jcameron** [comment]
Thanks!

**Comment by iliajie** [comment]
Jamie, I have revised this over again and I'm afraid the latest change isn't something ideal. As I was looking to find a bug based on this [report](https://forum.virtualmin.com/t/error-with-e-mail-users-after-moving-domain-to-new-server/125306?u=ilia), I noticed that after **Primary email address enabled** option is set to **No**:

  1. We don't disable neither **Forward to other addresses** or **Send automatic reply** options on the initial form submit
  2. Later, if a user is saved over again, the forward and automatic reply in `/etc/aliases` is getting removed as never sent to `save_user.cgi` file on submit.

This is the reason why hidden fields existed. We didn't show it in the UI but still re-submitted forward and auto-reply data on form re-saves, even if primary email option was disabled.

So, should we disable forward and auto-reply (i.e. remove a record from `/etc/aliases`) in the first time, i.e. at the moment when **Primary email address enabled** is set to **No** or we shouldn't touch forwards and auto-reply at all, keeping it preserved/intact?

**Comment by jcameron** [comment]
Can't we instead just detect this case in save_user.cgi and not update forwards or autoreplies? That seems a lot cleaner than having hidden HTML fields to pass along settings ...

**Comment by iliajie** [comment]
We could do it in `save_user.cgi` too. The question is do we really need to preserve it?

**Comment by jcameron** [comment]
In my opinion, if email is disabled all aliases should be removed .. the autoreply file should continue to exist, but actual autoreplies should stop working.

**Comment by jcameron** [comment]
Great thanks! So this gets rid of all the field hiding?

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
